File: backend-python/app/routers/themes.py
import time
import uuid

from fastapi import APIRouter, Body, HTTPException

from ..db import get_db
from ..deps import UserId

router = APIRouter()


def serialize(row) -> dict:
    return {"id": row["id"], "name": row["name"], "bg": row["bg"], "a": row["a"], "b": row["b"], "c": row["c"]}


@router.get("/")
def list_themes(user_id: UserId):
    with get_db() as db:
        rows = db.execute(
            "SELECT * FROM custom_themes WHERE user_id = ? ORDER BY created_at ASC",
            (user_id,),
        ).fetchall()
    return {"themes": [serialize(r) for r in rows]}


@router.post("/", status_code=201)
def create_theme(user_id: UserId, body: dict = Body(default={})):
    name = body.get("name")
    bg, a, b, c = body.get("bg"), body.get("a"), body.get("b"), body.get("c")
    if not bg or not a or not b or not c:
        raise HTTPException(status_code=400, detail="bg, a, b and c colors are required")
    theme_id = f"custom-{uuid.uuid4()}"
    with get_db() as db:
        db.execute(
            "INSERT INTO custom_themes (id, user_id, name, bg, a, b, c, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            (theme_id, user_id, name or "Custom theme", bg, a, b, c, int(time.time() * 1000)),
        )
    return {"theme": {"id": theme_id, "name": name or "Custom theme", "bg": bg, "a": a, "b": b, "c": c}}


@router.delete("/{theme_id}", status_code=204)
def delete_theme(theme_id: str, user_id: UserId):
    with get_db() as db:
        cur = db.execute(
            "DELETE FROM custom_themes WHERE id = ? AND user_id = ?", (theme_id, user_id)
        )
        if cur.rowcount == 0:
            raise HTTPException(status_code=404, detail="Theme not found")

File: backend/src/routes/themes.js
const express = require('express');
const { v4: uuid } = require('uuid');
const db = require('../db');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();
router.use(requireAuth);

function serialize(row) {
  return { id: row.id, name: row.name, bg: row.bg, a: row.a, b: row.b, c: row.c };
}

router.get('/', (req, res) => {
  const rows = db
    .prepare('SELECT * FROM custom_themes WHERE user_id = ? ORDER BY created_at ASC')
    .all(req.userId);
  res.json({ themes: rows.map(serialize) });
});

router.post('/', (req, res) => {
  const { name, bg, a, b, c } = req.body || {};
  if (!bg || !a || !b || !c) {
    return res.status(400).json({ error: 'bg, a, b and c colors are required' });
  }
  const id = `custom-${uuid()}`;
  db.prepare(
    'INSERT INTO custom_themes (id, user_id, name, bg, a, b, c, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)'
  ).run(id, req.userId, name || 'Custom theme', bg, a, b, c, Date.now());
  res.status(201).json({ theme: { id, name: name || 'Custom theme', bg, a, b, c } });
});

router.delete('/:id', (req, res) => {
  const result = db
    .prepare('DELETE FROM custom_themes WHERE id = ? AND user_id = ?')
    .run(req.params.id, req.userId);
  if (result.changes === 0) return res.status(404).json({ error: 'Theme not found' });
  res.status(204).end();
});

module.exports = router;
